<h1>Feature Engineering </h1>

In [77]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 

In [78]:
df=pd.read_csv('survey_results_cleaned.csv')
df.head()

respondent_id  age gender   zone            occupation income_levels  \
0        R00001   30      M  Urban  Working Professional          <10L   
1        R00002   46      F  Metro  Working Professional         > 35L   
2        R00003   41      F  Rural  Working Professional         > 35L   
3        R00004   33      F  Urban  Working Professional     16L - 25L   
4        R00005   23      M  Metro               Student  Not Reported   

  consume_frequency(weekly) current_brand preferable_consumption_size  \
0                 3-4 times      Newcomer             Medium (500 ml)   
1                 5-7 times   Established             Medium (500 ml)   
2                 3-4 times      Newcomer             Medium (500 ml)   
3                 5-7 times      Newcomer             Medium (500 ml)   
4                 3-4 times   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range  
0       Active (eg. Sports, gym)     100-150  
1           Social (eg. Parties)     200-250  
2       Active (eg. Sports, gym)     200-250  
3       Active (eg. Sports, gym)     150-200  
4       Active (eg. Sports, gym)      50-100

In [79]:
#creating the new columns 
df

respondent_id  age gender   zone            occupation income_levels  \
0            R00001   30      M  Urban  Working Professional          <10L   
1            R00002   46      F  Metro  Working Professional         > 35L   
2            R00003   41      F  Rural  Working Professional         > 35L   
3            R00004   33      F  Urban  Working Professional     16L - 25L   
4            R00005   23      M  Metro               Student  Not Reported   
...             ...  ...    ...    ...                   ...           ...   
29986        R29996   29      M  Metro  Working Professional     16L - 25L   
29987        R29997   42      F  Urban  Working Professional         > 35L   
29988        R29998   57      F  Metro               Retired          <10L   
29989        R29999   23      M  Urban  Working Professional     16L - 25L   
29990        R30000   26      M  Metro               Student  Not Reported   

      consume_frequency(weekly) current_brand preferable_consumption_size  \
0                     3-4 times      Newcomer             Medium (500 ml)   
1                     5-7 times   Established             Medium (500 ml)   
2                     3-4 times      Newcomer             Medium (500 ml)   
3                     5-7 times      Newcomer             Medium (500 ml)   
4                     3-4 times   Established             Medium (500 ml)   
...                         ...           ...                         ...   
29986                 3-4 times      Newcomer             Medium (500 ml)   
29987                 3-4 times   Established              Small (250 ml)   
29988                 0-2 times   Established             Medium (500 ml)   
29989                 5-7 times   Established              Small (250 ml)   
29990                 5-7 times      Newcomer             Medium (500 ml)   

      awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                        0 to 1                       Price       Traditional   
1                        2 to 4                     Quality            Exotic   
2                        2 to 4                Availability       Traditional   
3                        0 to 1            Brand Reputation            Exotic   
4                        0 to 1                Availability       Traditional   
...                         ...                         ...               ...   
29986                    2 to 4                Availability       Traditional   
29987                    2 to 4                     Quality       Traditional   
29988                    0 to 1                       Price       Traditional   
29989                   above 4                     Quality            Exotic   
29990                   above 4                       Price       Traditional   

      purchase_channel packaging_preference  \
0               Online               Simple   
1         Retail Store              Premium   
2         Retail Store              Premium   
3               Online         Eco-Friendly   
4               Online              Premium   
...                ...                  ...   
29986           Online               Simple   
29987           Online              Premium   
29988     Retail Store              Premium   
29989           Online         Eco-Friendly   
29990     Retail Store               Simple   

                            health_concerns typical_consumption_situations  \
0      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
1      Medium (Moderately health-conscious)           Social (eg. Parties)   
2      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
3                  Low (Not very concerned)       Active (eg. Sports, gym)   
4      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
...                                     ...                            ...   
29986          High (Very health-conscious)           Social

In [80]:
df.columns

Index(['respondent_id', 'age', 'gender', 'zone', 'occupation', 'income_levels',
       'consume_frequency(weekly)', 'current_brand',
       'preferable_consumption_size', 'awareness_of_other_brands',
       'reasons_for_choosing_brands', 'flavor_preference', 'purchase_channel',
       'packaging_preference', 'health_concerns',
       'typical_consumption_situations', 'price_range'],
      dtype='object')

In [81]:
# Define age groups
def categorize_age(age):
    if 18 <= age <= 25:
        return '18-25'
    elif 26 <= age <= 35:
        return '26-35'
    elif 36 <= age <= 45:
        return '36-45'
    elif 46 <= age <= 55:
        return '46-55'
    elif 56 <= age <= 70:
        return '56-70'
    else:
        return '70+'

# Add age_group column
df['age_group'] = df['age'].apply(categorize_age)

In [82]:
# df['age_group']=df.apply(age_group_classifier,axis=1)

In [83]:
df['age_group'].head()

0    26-35
1    46-55
2    36-45
3    26-35
4    18-25
Name: age_group, dtype: object

In [84]:
df.drop('age',axis=1,inplace=True)


In [85]:
df.columns

Index(['respondent_id', 'gender', 'zone', 'occupation', 'income_levels',
       'consume_frequency(weekly)', 'current_brand',
       'preferable_consumption_size', 'awareness_of_other_brands',
       'reasons_for_choosing_brands', 'flavor_preference', 'purchase_channel',
       'packaging_preference', 'health_concerns',
       'typical_consumption_situations', 'price_range', 'age_group'],
      dtype='object')

In [86]:
def consume_frequency_map(df):
    if df['consume_frequency(weekly)']=='0-2 times':
        return 1
    elif df['consume_frequency(weekly)']=='3-4 times':
        return 2
    elif df['consume_frequency(weekly)']=='5-7 times':
        return 3

In [87]:
df['consume_frequency(weekly)']=df.apply(consume_frequency_map,axis=1)

In [88]:
df['awareness_of_other_brands'].value_counts()

awareness_of_other_brands
0 to 1     12966
2 to 4     11225
above 4     5800
Name: count, dtype: int64

In [89]:
df.head()

respondent_id gender   zone            occupation income_levels  \
0        R00001      M  Urban  Working Professional          <10L   
1        R00002      F  Metro  Working Professional         > 35L   
2        R00003      F  Rural  Working Professional         > 35L   
3        R00004      F  Urban  Working Professional     16L - 25L   
4        R00005      M  Metro               Student  Not Reported   

   consume_frequency(weekly) current_brand preferable_consumption_size  \
0                          2      Newcomer             Medium (500 ml)   
1                          3   Established             Medium (500 ml)   
2                          2      Newcomer             Medium (500 ml)   
3                          3      Newcomer             Medium (500 ml)   
4                          2   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range age_group  
0       Active (eg. Sports, gym)     100-150     26-35  
1           Social (eg. Parties)     200-250     46-55  
2       Active (eg. Sports, gym)     200-250     36-45  
3       Active (eg. Sports, gym)     150-200     26-35  
4       Active (eg. Sports, gym)      50-100     18-25

In [90]:
# Function to map values
def awareness_map_function(value):
    if value == "0 to 1":
        return 1
    elif value == "2 to 4":
        return 2
    elif value == "above 4":
        return 3
    else:
        return None  # For any unmatched case

# Apply the function to the 'awareness_of_other_brands' column directly
df['awareness_of_other_brands_mapped'] = df['awareness_of_other_brands'].apply(awareness_map_function)


In [91]:
df.head()

respondent_id gender   zone            occupation income_levels  \
0        R00001      M  Urban  Working Professional          <10L   
1        R00002      F  Metro  Working Professional         > 35L   
2        R00003      F  Rural  Working Professional         > 35L   
3        R00004      F  Urban  Working Professional     16L - 25L   
4        R00005      M  Metro               Student  Not Reported   

   consume_frequency(weekly) current_brand preferable_consumption_size  \
0                          2      Newcomer             Medium (500 ml)   
1                          3   Established             Medium (500 ml)   
2                          2      Newcomer             Medium (500 ml)   
3                          3      Newcomer             Medium (500 ml)   
4                          2   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range age_group  \
0       Active (eg. Sports, gym)     100-150     26-35   
1           Social (eg. Parties)     200-250     46-55   
2       Active (eg. Sports, gym)     200-250     36-45   
3       Active (eg. Sports, gym)     150-200     26-35   
4       Active (eg. Sports, gym)      50-100     18-25   

   awareness_of_other_brands_mapped  
0                                 1  
1                                 2  
2                                 2  
3                                 1  
4                                 1

In [92]:
df['cf_ab_score']=df['consume_frequency(weekly)']/(df['consume_frequency(weekly)']+df['awareness_of_other_brands_mapped'])
df['cf_ab_score']=round(df['cf_ab_score'],2)

In [93]:
df['cf_ab_score'].max()

np.float64(0.75)

In [94]:
df.head()

respondent_id gender   zone            occupation income_levels  \
0        R00001      M  Urban  Working Professional          <10L   
1        R00002      F  Metro  Working Professional         > 35L   
2        R00003      F  Rural  Working Professional         > 35L   
3        R00004      F  Urban  Working Professional     16L - 25L   
4        R00005      M  Metro               Student  Not Reported   

   consume_frequency(weekly) current_brand preferable_consumption_size  \
0                          2      Newcomer             Medium (500 ml)   
1                          3   Established             Medium (500 ml)   
2                          2      Newcomer             Medium (500 ml)   
3                          3      Newcomer             Medium (500 ml)   
4                          2   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range age_group  \
0       Active (eg. Sports, gym)     100-150     26-35   
1           Social (eg. Parties)     200-250     46-55   
2       Active (eg. Sports, gym)     200-250     36-45   
3       Active (eg. Sports, gym)     150-200     26-35   
4       Active (eg. Sports, gym)      50-100     18-25   

   awareness_of_other_brands_mapped  cf_ab_score  
0                                 1         0.67  
1                                 2         0.60  
2                                 2         0.50  
3                                 1         0.75  
4                                 1         0.67

In [95]:
df['zone']=df['zone'].map({'Urban':3,"Metro":4,"Rural":1,"Semi-Urban":2})

In [96]:
df['income_levels']=df['income_levels'].map({"<10L":1,"10L - 15L":2, "16L - 25L":3,"26L - 35L":4,"> 35L":5,"Not Reported" :0})

In [97]:
df

respondent_id gender  zone            occupation  income_levels  \
0            R00001      M     3  Working Professional              1   
1            R00002      F     4  Working Professional              5   
2            R00003      F     1  Working Professional              5   
3            R00004      F     3  Working Professional              3   
4            R00005      M     4               Student              0   
...             ...    ...   ...                   ...            ...   
29986        R29996      M     4  Working Professional              3   
29987        R29997      F     3  Working Professional              5   
29988        R29998      F     4               Retired              1   
29989        R29999      M     3  Working Professional              3   
29990        R30000      M     4               Student              0   

       consume_frequency(weekly) current_brand preferable_consumption_size  \
0                              2      Newcomer             Medium (500 ml)   
1                              3   Established             Medium (500 ml)   
2                              2      Newcomer             Medium (500 ml)   
3                              3      Newcomer             Medium (500 ml)   
4                              2   Established             Medium (500 ml)   
...                          ...           ...                         ...   
29986                          2      Newcomer             Medium (500 ml)   
29987                          2   Established              Small (250 ml)   
29988                          1   Established             Medium (500 ml)   
29989                          3   Established              Small (250 ml)   
29990                          3      Newcomer             Medium (500 ml)   

      awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                        0 to 1                       Price       Traditional   
1                        2 to 4                     Quality            Exotic   
2                        2 to 4                Availability       Traditional   
3                        0 to 1            Brand Reputation            Exotic   
4                        0 to 1                Availability       Traditional   
...                         ...                         ...               ...   
29986                    2 to 4                Availability       Traditional   
29987                    2 to 4                     Quality       Traditional   
29988                    0 to 1                       Price       Traditional   
29989                   above 4                     Quality            Exotic   
29990                   above 4                       Price       Traditional   

      purchase_channel packaging_preference  \
0               Online               Simple   
1         Retail Store              Premium   
2         Retail Store              Premium   
3               Online         Eco-Friendly   
4               Online              Premium   
...                ...                  ...   
29986           Online               Simple   
29987           Online              Premium   
29988     Retail Store              Premium   
29989           Online         Eco-Friendly   
29990     Retail Store               Simple   

                            health_concerns typical_consumption_situations  \
0      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
1      Medium (Moderately health-conscious)           Social (eg. Parties)   
2      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
3                  Low (Not very concerned)       Active (eg. Sports, gym)   
4      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
...                                     ...                            ...   
29986          High (Very health-conscious)           Social (eg. Parties)   
29987              Low (Not ve

In [98]:
df['zas_score']=df['zone']*df['income_levels']


In [99]:
df['zas_score'].unique()

array([ 3, 20,  5,  9,  0,  6, 12,  2,  4,  8, 15, 16,  1, 10])

In [100]:
df.head()

respondent_id gender  zone            occupation  income_levels  \
0        R00001      M     3  Working Professional              1   
1        R00002      F     4  Working Professional              5   
2        R00003      F     1  Working Professional              5   
3        R00004      F     3  Working Professional              3   
4        R00005      M     4               Student              0   

   consume_frequency(weekly) current_brand preferable_consumption_size  \
0                          2      Newcomer             Medium (500 ml)   
1                          3   Established             Medium (500 ml)   
2                          2      Newcomer             Medium (500 ml)   
3                          3      Newcomer             Medium (500 ml)   
4                          2   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range age_group  \
0       Active (eg. Sports, gym)     100-150     26-35   
1           Social (eg. Parties)     200-250     46-55   
2       Active (eg. Sports, gym)     200-250     36-45   
3       Active (eg. Sports, gym)     150-200     26-35   
4       Active (eg. Sports, gym)      50-100     18-25   

   awareness_of_other_brands_mapped  cf_ab_score  zas_score  
0                                 1         0.67          3  
1                                 2         0.60         20  
2                                 2         0.50          5  
3                                 1         0.75          9  
4                                 1         0.67          0

In [101]:
df['current_brand'].unique()

array(['Newcomer', 'Established'], dtype=object)

In [102]:
# df_filtered['bsi'] = ((df_filtered['current_brand'] != "Established") &
#                       (df_filtered['reasons_for_choosing_brands'].isin(["Price", "Quality"]))).astype(int)

# # Apply row-wise
# df['bsi'] = df.apply(bsi_calculate, axis=1)


In [103]:
# df['bsi']=df.apply(bsi_calculate,axis=1)

In [104]:
df.head()

respondent_id gender  zone            occupation  income_levels  \
0        R00001      M     3  Working Professional              1   
1        R00002      F     4  Working Professional              5   
2        R00003      F     1  Working Professional              5   
3        R00004      F     3  Working Professional              3   
4        R00005      M     4               Student              0   

   consume_frequency(weekly) current_brand preferable_consumption_size  \
0                          2      Newcomer             Medium (500 ml)   
1                          3   Established             Medium (500 ml)   
2                          2      Newcomer             Medium (500 ml)   
3                          3      Newcomer             Medium (500 ml)   
4                          2   Established             Medium (500 ml)   

  awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                    0 to 1                       Price       Traditional   
1                    2 to 4                     Quality            Exotic   
2                    2 to 4                Availability       Traditional   
3                    0 to 1            Brand Reputation            Exotic   
4                    0 to 1                Availability       Traditional   

  purchase_channel packaging_preference                       health_concerns  \
0           Online               Simple  Medium (Moderately health-conscious)   
1     Retail Store              Premium  Medium (Moderately health-conscious)   
2     Retail Store              Premium  Medium (Moderately health-conscious)   
3           Online         Eco-Friendly              Low (Not very concerned)   
4           Online              Premium  Medium (Moderately health-conscious)   

  typical_consumption_situations price_range age_group  \
0       Active (eg. Sports, gym)     100-150     26-35   
1           Social (eg. Parties)     200-250     46-55   
2       Active (eg. Sports, gym)     200-250     36-45   
3       Active (eg. Sports, gym)     150-200     26-35   
4       Active (eg. Sports, gym)      50-100     18-25   

   awareness_of_other_brands_mapped  cf_ab_score  zas_score  
0                                 1         0.67          3  
1                                 2         0.60         20  
2                                 2         0.50          5  
3                                 1         0.75          9  
4                                 1         0.67          0

In [105]:
df['zas_score'].describe()

count    29991.000000
mean         6.096529
std          5.517959
min          0.000000
25%          0.000000
50%          6.000000
75%          9.000000
max         20.000000
Name: zas_score, dtype: float64

In [106]:
df.columns

Index(['respondent_id', 'gender', 'zone', 'occupation', 'income_levels',
       'consume_frequency(weekly)', 'current_brand',
       'preferable_consumption_size', 'awareness_of_other_brands',
       'reasons_for_choosing_brands', 'flavor_preference', 'purchase_channel',
       'packaging_preference', 'health_concerns',
       'typical_consumption_situations', 'price_range', 'age_group',
       'awareness_of_other_brands_mapped', 'cf_ab_score', 'zas_score'],
      dtype='object')

In [107]:
df.groupby('occupation')['age_group'].value_counts().unstack()

age_group              18-25   26-35   36-45   46-55   56-70
occupation                                                  
Entrepreneur           535.0  1826.0  1619.0   799.0   221.0
Retired                  NaN     NaN     NaN     NaN  1130.0
Student               7328.0   697.0     NaN     NaN    35.0
Working Professional  2605.0  6570.0  4353.0  2167.0   106.0

In [108]:
df

respondent_id gender  zone            occupation  income_levels  \
0            R00001      M     3  Working Professional              1   
1            R00002      F     4  Working Professional              5   
2            R00003      F     1  Working Professional              5   
3            R00004      F     3  Working Professional              3   
4            R00005      M     4               Student              0   
...             ...    ...   ...                   ...            ...   
29986        R29996      M     4  Working Professional              3   
29987        R29997      F     3  Working Professional              5   
29988        R29998      F     4               Retired              1   
29989        R29999      M     3  Working Professional              3   
29990        R30000      M     4               Student              0   

       consume_frequency(weekly) current_brand preferable_consumption_size  \
0                              2      Newcomer             Medium (500 ml)   
1                              3   Established             Medium (500 ml)   
2                              2      Newcomer             Medium (500 ml)   
3                              3      Newcomer             Medium (500 ml)   
4                              2   Established             Medium (500 ml)   
...                          ...           ...                         ...   
29986                          2      Newcomer             Medium (500 ml)   
29987                          2   Established              Small (250 ml)   
29988                          1   Established             Medium (500 ml)   
29989                          3   Established              Small (250 ml)   
29990                          3      Newcomer             Medium (500 ml)   

      awareness_of_other_brands reasons_for_choosing_brands flavor_preference  \
0                        0 to 1                       Price       Traditional   
1                        2 to 4                     Quality            Exotic   
2                        2 to 4                Availability       Traditional   
3                        0 to 1            Brand Reputation            Exotic   
4                        0 to 1                Availability       Traditional   
...                         ...                         ...               ...   
29986                    2 to 4                Availability       Traditional   
29987                    2 to 4                     Quality       Traditional   
29988                    0 to 1                       Price       Traditional   
29989                   above 4                     Quality            Exotic   
29990                   above 4                       Price       Traditional   

      purchase_channel packaging_preference  \
0               Online               Simple   
1         Retail Store              Premium   
2         Retail Store              Premium   
3               Online         Eco-Friendly   
4               Online              Premium   
...                ...                  ...   
29986           Online               Simple   
29987           Online              Premium   
29988     Retail Store              Premium   
29989           Online         Eco-Friendly   
29990     Retail Store               Simple   

                            health_concerns typical_consumption_situations  \
0      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
1      Medium (Moderately health-conscious)           Social (eg. Parties)   
2      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
3                  Low (Not very concerned)       Active (eg. Sports, gym)   
4      Medium (Moderately health-conscious)       Active (eg. Sports, gym)   
...                                     ...                            ...   
29986          High (Very health-conscious)           Social (eg. Parties)   
29987              Low (Not ve

In [109]:

df_filtered = df[~((df['age_group'] == '56-70') & (df['occupation'] == 'Student'))]

# Count remaining rows
remaining_rows = len(df_filtered)
print(f"Remaining rows: {remaining_rows}")


Remaining rows: 29956


In [110]:
# df['respondent_id'].describe()

In [111]:
# # Remove logical outliers
# df_filtered = df[~((df['age_group'] == '56-70') & (df['occupation'] == 'Student'))]
# # Count remaining rows
# remaining_rows = len(df_filtered)
# print(f"Remaining rows: {remaining_rows}")

Perfect !!

We have cleaned the Outliers

In [112]:

df_filtered['bsi'] = ((df_filtered['current_brand'] != "Established") &
                      (df_filtered['reasons_for_choosing_brands'].isin(["Price", "Quality"]))).astype(int)
# Count rows where BSI = 0
bsi_zero_count = (df_filtered['bsi'] == 0).sum()

print(f"Number of rows with BSI = 0: {bsi_zero_count}")

Number of rows with BSI = 0: 20796


C:\Users\hp\AppData\Local\Temp\ipykernel_26684\2811497589.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['bsi'] = ((df_filtered['current_brand'] != "Established") &


The Number of rows with BSI after removing logical outliers is = 20796

In [114]:
df_filtered.to_csv("Survey_data_Ready_For_Modelling.csv",index=False)